pkg/controller/pkg/revision/reconciler.go

```diff
@@ -0,0 +1,398 @@
+		}
+		if err := r.client.Delete(ctx, s); resource.IgnoreNotFound(err) != nil {
+			log.Debug(errDeleteProviderSA, "error", err)
+			r.record.Event(pr, event.Warning(errDeleteProviderSA, err))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+		}
+	}
+
+	// NOTE(hasheddan): the following logic enforces the present guarantees of
+	// the package manager, which includes:
+	// 1. We will not create or update any objects unless we can create or
+	// update all objects
+	// 2. We will not create or update any objects unless we can gain control of
+	// all objects
+	// 3. We will only modify owner references when the package revision is
+	// inactive
+	// 4. If objects are reconciled by a packaged controller, we will not remove
+	// owner references until the controller has been stopped
+	if err := r.establisher.Check(ctx, pkg.GetObjects(), pr, pr.GetDesiredState() == v1alpha1.PackageRevisionActive); err != nil {
+		log.Debug(errEstablishControl, "error", err)
+		r.record.Event(pr, event.Warning(errEstablishControl, err))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+	}
+
+	if err := r.establisher.Establish(ctx, pr, pr.GetDesiredState() == v1alpha1.PackageRevisionActive); err != nil {
+		log.Debug(errEstablishControl, "error", err)
+		r.record.Event(pr, event.Warning(errEstablishControl, err))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+	}
+
+	// If package is a provider package, all resources have been created, and
+	// the revision is active, we should start the controller.
+	if pr.GetObjectKind().GroupVersionKind() == v1alpha1.ProviderRevisionGroupVersionKind && desiredState == v1alpha1.PackageRevisionActive {
+		// TODO(hasheddan): see comment on conditional logic above.
+		pkgProvider := pkg.GetMeta()[0].(*pkgmeta.Provider)
+		s, d := buildProviderDeployment(pkgProvider, pr, r.namespace)
+		if err := r.client.Applicator.Apply(ctx, s); err != nil {
+			log.Debug(errCreateProviderSA, "error", err)
+			r.record.Event(pr, event.Warning(errCreateProviderSA, err))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+		}
+		if err := r.client.Applicator.Apply(ctx, d); err != nil {
+			log.Debug(errCreateProviderDeployment, "error", err)
+			r.record.Event(pr, event.Warning(errCreateProviderDeployment, err))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+		}
+		pr.SetControllerReference(&runtimev1alpha1.Reference{Name: d.GetName()})
+	}
+
+	// Update object list in package revision status and set ready condition to
+	// true.
+	pr.SetObjects(r.establisher.GetResourceRefs())
+	// TODO(hasheddan): set remaining status fields (dependencies, crossplane
+	// version, permission requests)
```

Comment:
Does this TODO need addressing before we merge, or do you want to handle it in a future PR?

Reply:
Future PR :+1: Related to how RBAC controller will create roles for controllers